### DISCUSSION TITLE
How MinIO find the storage pool of a object

### DISCUSSION THREAD

**Discussion by yidong-pony-ai** [minio/minio#13443] [Q&A] [answered]
Hi Guys,
I'm reading the "Distributed Server Design Guide" and still a bit confused about the storage pool management.
The doc says:
> MinIO also supports expansion of existing clusters in server pools. Each pool is a self contained entity with same SLA's (read/write quorum) for each object as original cluster. **_By using the existing namespace for lookup validation_** MinIO ensures conflicting objects are not created. When no such object exists then MinIO simply uses the least used pool to place new objects.

Is there a large table to maintain the object ->storagePool mapping for lookup? Or it create multiple queries to each storage pool for lookup?

If the latter one I assume:
- I assume the read path also follow the same way? Say I've 10 storage pool, then each object read will create10 queries to each storage pool so bad for performance to have too many pools?
- What happen if a storage pool is completely down? Seems it won't block new object creation on other pools. Then does it mean it may create a conflicting name object as not possible to query the failed storage pool?

**Comment by klauspost** [comment, marked as answer]
> I assume the read path also follow the same way? Say I've 10 storage pool, then each object read will create10 queries to each storage pool so bad for performance to have too many pools?

Correct. We generally recommend having a forward looking provisioning plan and not end up with 10 server pools.

> What happen if a storage pool is completely down? Seems it won't block new object creation on other pools. Then does it mean it may create a conflicting name object as not possible to query the failed storage pool?

Correct. Multiple returned objects are sorted by modification dates, example: https://github.com/minio/minio/blob/ec0fee620843cf8f878d19a00d44c016cc2323ab/cmd/erasure-server-pool.go#L311
